### DISCUSSION TITLE
libcurl share with multiple threads bug questions

### DISCUSSION THREAD

**Discussion by jbaldwin** [curl/curl#6893] [Q&A] [answered]
Hello, I am attempting to look into this #4915 as it has some very large wins for our application around open TCP sockets, roughly -50%.  I've generally managed to fix what I think are the low hanging fruit crashes via TSAN using the "data connect" mutex (`CONNCACHE_LOCK` and `CONNCACHE_UNLOCK`) in appropriate locations and running against some of our production load to identify other locations.  This bug generally causes our application to crash within 30 minutes with no code changes on live traffic using the latest version of libcurl.  Our app does roughly 5k to 10k http requests/sec/server to give some scale on how much traffic is needed to trigger the bug(s).

(1) Low hanging fruit data locks (no questions here, just posting for visibility)

https://github.com/jbaldwin/curl/commit/f2d991ba792dc7faa0800d93e5360c436e6ed041

(2) I'm wondering about is how to best resolve this issue that TSAN identified:

https://github.com/jbaldwin/curl/commit/a814d3db8254f24a0f98d51d713ae0f1de549deb

I do not see any atomic operations in the libcurl codebase and the solution I have here is obviously gcc + linux specific, does libcurl have a specific way to handle atomic compare and swap operations?  The other option I was thinking here would be to call the SSL user mutex locking functions, but that seems like a lot of overhead for a simple call once function.

(3) Double locking critical section identified

https://github.com/jbaldwin/curl/commit/679e788f9e2b145d5046c8466558d8b11ee6ef8d

This section appears to be a critical section that requires the SSL lock as well as the DATA lock, I haven't run into a deadlock (yet) but it concerns me that this section needs both, and its not apparent to me why yet since these are only reads on the easy handle data structure.  But perhaps another thread has access to the easy handle and is writing to them while they are being read here.  Is this solution in your mind acceptable?

I also think this area might want to hoist the data lock up a level since it does look like there is another function of similar kind that is also reading the easy handles connection data, but that will require editing every ssl file type (that is OK, but I was specifically just trying to fix my use case of OpenSSL first).  Those functions from what I can tell are like the following, but possibly more:

`Curl_ssl_getsessionid, Curl_ssl_delsessionid, Curl_ssl_addsessionid`

(4) Testing this?  Considering how much volume/throughput our app requires to trigger is this worth adding into the curl test suite in some capacity assuming these changes do ultimately fix the issue?

If this is the wrong place to ask these questions let me know, happy to move the discussion elsewhere.

Thanks,
Josh

**Comment by bagder** [comment, marked as answer]
(you would probably reach more libcurl hackers on the curl-library mailing list)

> (1) Low hanging fruit data locks (no questions here, just posting for visibility)

I would like to have a better grasp on exactly why you need to mutex the `conn->bits` accesses. My guess is that since some of them are used by `url.c:ConnectionExists()` and they are bitfields, the "neighbor bits" in the bitfields aren't threadsafe? Then maybe those that really need to be accessed from multiple threads should be moved to a separate `bits` field so that we can make sure to always mutex those accesses and avoid mutexing bits accesses that don't need it...

> (2) I do not see any atomic operations in the libcurl codebase

There is none. libcurl is C89 and can be used without a thread library so there's no such thing universally available. Hence we use the mutex callbacks even for that.

> (3) Double locking critical section identified

Ouch. But yeah, I suppose rearranging code to avoid it is the only proper answer I can give.

> (4) Testing this

That's really tricky. But I think doing a test case that runs for X seconds using Y threads and we use TSAN or valgrind, we can see if we can come up with something that A) fails before any fix and B) runs fine after the fixes. Races and thread-safety are hard things to test but let's try to do it as good as we can!

**Reply by jbaldwin** [reply]
Thanks for the quick reply!  I'll do some more work based on your feedback here and then re-post to the mailing list.  Just some quick thoughts though:

(1) Yeah I'm not that familiar with each field in the connection bits section (this is really my first time diving really deep into the curl code) but TSAN identified these as having concurrent reads/writes on different threads so I've placed locks around them for now, I do like the idea of identifying them as more granular, I will try and look into this more.

(2) Cool, this makes sense, I will switch it to use the mutex, thanks for the explanation and decision around not using atomics.

(3) I did run into a deadlock after running it over night, frames 17 and 0 are trying to lock the same `CONCACHE_LOCK`, I think a recursive mutex would solve this since its on the same thread but I'm not sure that is an acceptable solution for the feature to require users to provide recursive mutexes in the callbacks.

```
#0  0x0000000000951057 in Curl_ssl_getsessionid ()  <-- CONCACHE_LOCK()
#1  0x00000000009534b8 in ossl_new_session_cb ()
#2  0x000000000096aede in ssl_update_cache ()
#3  0x00000000009b7580 in tls_process_new_session_ticket ()
#4  0x00000000009b3475 in ossl_statem_client_process_message ()
#5  0x000000000096ff0a in read_state_machine ()
#6  0x000000000096f9a2 in state_machine ()
#7  0x000000000096f4a8 in ossl_statem_connect ()
#8  0x0000000000977bf9 in ssl3_read_bytes ()
#9  0x000000000096276d in ssl3_read_internal ()
#10 0x0000000000962848 in ssl3_read ()
#11 0x00000000009677c1 in ssl_read_internal ()
#12 0x000000000096781e in SSL_read ()
#13 0x0000000000958c3f in ossl_recv ()
#14 0x00000000008fa3f4 in http2_conncheck ()
#15 0x000000000092ed3b in extract_if_dead.part.0 ()
#16 0x000000000092ede7 in call_extract_if_dead ()
#17 0x00000000008f0045 in Curl_conncache_foreach () <-- CONCACHE_LOCK()
#18 0x0000000000933979 in Curl_connect ()
#19 0x0000000000913964 in multi_runsingle ()
#20 0x0000000000915590 in multi_socket ()
#21 0x0000000000915844 in curl_multi_socket_action ()
#22 <APP_CODE>
```

(4) Alright, I'll give this some thought as well.  Sort of inline with what I was thinking, but it might be difficult getting it to reliably reproduce on the segfaults for part (A).
